internal/pkg/pluginengine/change_helper.go

```diff
@@ -59,49 +59,60 @@ func drifted(a, b map[string]interface{}) bool {
 func changesForApply(smgr statemanager.Manager, cfg *configloader.Config) ([]*Change, error) {
 	changes := make([]*Change, 0)
 
-	// 1, create a temporary state map used to store unprocessed tools.
+	// 1. create a temporary state map used to store unprocessed tools.
 	tmpStatesMap := smgr.GetStatesMap().DeepCopy()
 
-	// 2, for each tool in the config, generate changes.
-	for _, tool := range cfg.Tools {
-		state := smgr.GetState(statemanager.StateKeyGenerateFunc(&tool))
+	// 2. handle dependency and sort the tools in the config
+	var tools [][]configloader.Tool
```

Comment:
The tools with type `[][]configloader.Tool` and the tools with type `[]configloader.Tool` in `topologicalSort(tools []configloader.Tool)` confuse me a bit. Same name but different type here. Maybe there's a better way to name it